java-frontend/src/main/java/org/sonar/java/ast/visitors/FileLinesVisitor.java

```diff
@@ -19,63 +19,191 @@
 
-  public FileLinesVisitor(SonarComponents sonarComponents, Charset charset) {
+  public FileLinesVisitor(SonarComponents sonarComponents) {
     this.sonarComponents = sonarComponents;
-    this.charset = charset;
   }
 
   @Override
   public List<Tree.Kind> nodesToVisit() {
-    return ImmutableList.of(Tree.Kind.TOKEN);
+    if(sonarComponents.isSQGreaterThan62()) {
+      return ImmutableList.of(TOKEN,
+        METHOD, CONSTRUCTOR,
+        INITIALIZER, STATIC_INITIALIZER,
+        VARIABLE,
+        FOR_EACH_STATEMENT, FOR_STATEMENT, WHILE_STATEMENT, DO_STATEMENT,
+        LAMBDA_EXPRESSION);
+    }
+    return ImmutableList.of(TOKEN);
   }
 
   @Override
   public void scanFile(JavaFileScannerContext context) {
     super.scanFile(context);
-
-    FileLinesContext fileLinesContext = sonarComponents.fileLinesContextFor(context.getFile());
-    // TODO minimize access to files, another one in LinesVisitor
-    int fileLength;
-    try {
-      fileLength = Files.readLines(context.getFile(), charset).size();
-    } catch (IOException e) {
-      throw Throwables.propagate(e);
-    }
+    File currentFile = context.getFile();
+    FileLinesContext fileLinesContext = sonarComponents.fileLinesContextFor(currentFile);
+    int fileLength = sonarComponents.fileLength(currentFile);
     for (int line = 1; line <= fileLength; line++) {
       fileLinesContext.setIntValue(CoreMetrics.NCLOC_DATA_KEY, line, linesOfCode.contains(line) ? 1 : 0);
       fileLinesContext.setIntValue(CoreMetrics.COMMENT_LINES_DATA_KEY, line, linesOfComments.contains(line) ? 1 : 0);
+      if(sonarComponents.isSQGreaterThan62()) {
+        fileLinesContext.setIntValue(CoreMetrics.EXECUTABLE_LINES_DATA_KEY, line, executableLines.contains(line) ? 1 : 0);
+      }
     }
     fileLinesContext.save();
 
     linesOfCode.clear();
     linesOfComments.clear();
+    executableLines.clear();
+  }
+
+  @Override
+  public void visitNode(Tree tree) {
+    List<? extends Tree> trees = Collections.emptyList();
+    switch (tree.kind()) {
+      case INITIALIZER:
+      case STATIC_INITIALIZER:
+        trees = ((BlockTree) tree).body();
+        break;
+      case VARIABLE:
+        VariableTree variableTree = (VariableTree) tree;
+        trees = visitVariable(variableTree);
+        break;
+      case LAMBDA_EXPRESSION:
+        LambdaExpressionTree lambda = (LambdaExpressionTree) tree;
+        trees = visitLambda(lambda);
+        break;
+      case METHOD:
+      case CONSTRUCTOR:
+        trees = visitMethod((MethodTree) tree);
+        break;
+      case FOR_STATEMENT:
+      case FOR_EACH_STATEMENT:
+      case WHILE_STATEMENT:
+      case DO_STATEMENT:
+        executableLines.add(tree.lastToken().line());
+        break;
+    }
+    computeExecutableLines(trees);
+  }
+
+  private List<? extends Tree> visitVariable(VariableTree variableTree) {
+    ExpressionTree initializer = variableTree.initializer();
+    if(initializer != null && !isConstant(variableTree)) {
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) Reduce this switch case number of lines from 10 to at most 5, for example by extracting code into methods. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS1151)